docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,212 @@
+
+For more information on the various approaches and proposals, please see the
+[mempool discussion](https://github.com/tendermint/tendermint/discussions/6295).
+
+## Decision
+
+To incorporate a priority-based flexible and performant mempool in Tendermint Core,
+we will introduce new fields, `priority` and `sender`, into the `ResponseCheckTx`
+type and augment the existing mempool data structure to facilitate prioritization
+of uncommitted transactions in addition to extended functionality such as
+replace-by-priority and allowing multiple transactions to exist from the same
+sender with varying priorities.
+
+## Detailed Design
+
+### CheckTx
+
+We introduce the following new fields into the `ResponseCheckTx` type:
+
+```diff
+message ResponseCheckTx {
+  uint32         code       = 1;
+  bytes          data       = 2;
+  string         log        = 3;  // nondeterministic
+  string         info       = 4;  // nondeterministic
+  int64          gas_wanted = 5 [json_name = "gas_wanted"];
+  int64          gas_used   = 6 [json_name = "gas_used"];
+  repeated Event events     = 7 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string         codespace  = 8;
++ int64          priority   = 9;
++ string         sender     = 10;
+}
+```
+
+It is entirely up the application in determining how these fields are populated
+and with what values, e.g. the `sender` could be the signer and fee payer 
+the transaction, the `priority` could be the cumulative sum of the fee(s).
+
+Only `sender` is required, while `priority` can be omitted which would result in
+using the default value of zero.
+
+### Mempool
+
+The existing concurrent-safe linked-list will be removed entirely in favor of a
+thread-safe map of `<sender:*Tx>`, i.e a mapping from `sender` to a single `*Tx`
+object, where each `*Tx` is the next valid and processable transaction from the
+given `sender`.
+
+On top of this mapping, we index all transactions by priority using a thread-safe
+priority queue, i.e. a [max heap](https://en.wikipedia.org/wiki/Min-max_heap).
+When a proposer is ready to select transactions for the next block proposal,
+transactions are selected from this priority index by highest priority order.
+When a transaction is selected and reaped, it is removed from this index and
+from the `<sender:*Tx>` mapping.
+
+We define `Tx` as the following data structure:
+
+```go
+type Tx struct {
+  // Tx represents the raw binary transaction data.
+  Tx []byte
+
+  // Priority defines the transaction's priority as specified by the application
+  // in the ResponseCheckTx response.
+  Priority int64
+
+  // Sender defines the transaction's sender as specified by the application in
+  // the ResponseCheckTx response.
+  Sender string
+
+  // Index defines the current index in the priority queue index. Note, if
+  // multiple Tx indexes are needed, this field will be removed and each Tx
+  // index will have its own wrapped Tx type.
+  Index int
```

Comment:
Might we also want to track `Gas` as an `int64`

Reply:
What do you mean exactly? We're not using the notion of gas in this refactor.

Reply:
Right this refactor might not alter the behavior of gas in any way but as were redefining the data structures within mempool, I thought we would want to be complete and thus `Gas` is needed to determine how many txs to include when calling `Reap`

Reply:
As far as I know, the mempool doesn't actually have any data structures? I could be wrong. But yes, we can totally add `Gas` to this wrapper type 👍